Share field mapper tests with other modules

shans96 opened the pull request:
WIP PR to close issue #92947

My approach for solving this problem was to see if I could get `UnsignedLongMapperTests` to extend `NumberFieldMapperTests` by copying `NumberFieldMapperTests` over to the test framework (and then do the same for `ScaledFloatFieldMapperTests`). Once I've made `UnsignedLongMapperTests` extend `NumberFieldMapperTests`, my plan was to make it extend `WholeNumberFieldMapperTests` instead, as that seems more appropriate (and specialised) for the unsigned long type.

I have managed to make `UnsignedLongMapperTests` extend  `NumberFieldMapperTests` but there are now some test failures for:
-`testCoerce`
-`testOutOfRangeValues`
-`testAllowMultipleValuesField`
-`testIgnoreMalformedWithObject`
-`testScriptableTypes`

It looks like some wider changes are needed to fix the tests (e.g., `NumberFieldMapper` has to be updated with some code to support unsigned longs), but I'm not sure of whether this is the right decision or not from an architectural perspective, so I wanted to get some early feedback to make sure I'm on the right track.

Any insight on the above would be appreciated!

shans96 committed:
Move IngestScriptSupport out of MapperTestCase into its own class so it can be reused in test classes

shans96 committed:
Copy NumberFieldMapperTests to test framework and use it in UnsignedLongFieldMapperTests

shans96 committed:
Fix typo and access modifier

shans96 committed:
Implement outOfRangeSpecs for UnsignedLongFieldMapperTests{

shans96 committed:
Implement missingValue

shans96 committed:
Partially implement randomNumber(), add TODO for it

shans96 committed:
Change comment to TODO

shans96 committed:
Clarify comment

shans96 committed:
Add changelog

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

cbuescher commented on server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:
```diff
@@ -336,12 +336,12 @@ protected <T> T compileOtherScript(Script script, ScriptContext<T> context) {
 
             @Override
-            ScriptFactory emptyFieldScript() {
+            protected ScriptFactory emptyFieldScript() {
```
This class wouldn't bee needed in the "server" module after the move I think. Small changes in visibility are okay if needed, please comment them if you have questions.

cbuescher reviewed (commented):
Hi @shans96, I took a quick look at this PR but didn't look into all the details because you still marked it as draft.
The general idea is right, I left some comments on some specifics.
I think the general ask with this issue is to move the testing code and put the two mentioned tests (UnsignedLongFieldMapperTests, ScaledFloatFieldMapperTests) under one of the more specific base tests. I think that when tests of the base class fail (e.g. "coerce" doesn't seem to be supported in UnsignedLongFieldMapper), we should overwrite or disable those tests in the inheriting test class. If only some assertions need to be fixed, that's okay of course, but I would like leave behavioural changes out of this refactoring PR as much as possible.
Hope that helps, let me know if you need more input.

shans96 committed:
Move NumberFieldMapperTests using git to preserve history

shans96 committed:
Revert "Move IngestScriptSupport out of MapperTestCase into its own class so it can be reused in test classes"

This reverts commit 33a9526b49a00a23de3a93bcab3e41153a4dd8c0.

shans96 committed:
Fix type mentioned in changelog

shans96 committed:
Override testCoerce

shans96 committed:
Return empty list of outOfRangeSpecs to get the test passing

shans96 committed:
Override testIgnoreMalformedWithObject to make it pass

shans96 committed:
Override testAllowMultipleValuesField to make it pass

shans96 committed:
Override testScriptableTypes to make it pass

shans96 committed:
Move WholeNumberFieldMapperTests to the test framework, use it in UnsignedLongFieldMapperTests

shans96 committed:
Make ScaledFloatFieldMapperTests extend NumberFieldMapperTests

shans96 committed:
Remove unused imports

shans96 commented on modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapperTests.java:
```diff
@@ -196,7 +197,7 @@ public void testStore() throws Exception {
     }
 
-    public void testCoerce() throws Exception {
+    public void testCoerce() throws IOException {
```
Changed to make the method signature match how it is declared in `NumberFieldMapperTests`

shans96 commented on test/framework/src/main/java/org/elasticsearch/index/mapper/IngestScriptSupport.java:
```diff
@@ -0,0 +1,49 @@
+
+// Copied from MapperTestCase
+// TODO: Is it fine to keep this class in the index package and keep the weaker access levels for it, or should it be moved somewhere else?
+public abstract class IngestScriptSupport {
```
Good point, thanks- don't think it was needed anymore

shans96 commented on test/framework/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:
```diff
@@ -315,7 +317,7 @@ public void testMetricAndDocvalues() {
 
     @Override
-    protected final Object generateRandomInputValue(MappedFieldType ft) {
+    protected Object generateRandomInputValue(MappedFieldType ft) {
```
Needs to be made non-final so `UnsignedLongFieldMapperTests` and `ScaledFloatFieldMapperTests` can override it

shans96 commented on modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapperTests.java:
```diff
@@ -432,6 +433,39 @@ protected IngestScriptSupport ingestScriptSupport() {
+    }
+
+    @Override
+    public void testIgnoreMalformedWithObject() {} // unimplemented
```
I've overridden the methods for now but I'm sure there's a cleaner way to write these. I've seen that some tests can be skipped in gradle, would that be cleaner, and if so, where should I look to find an example?

shans96 committed:
Reduce access levels

shans96 commented on test/framework/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:
```diff
@@ -459,4 +461,30 @@ public List<SyntheticSourceInvalidExample> invalidExample() throws IOException {
         }
     }
+
+    // Copied from NumberFieldTypeTests, better to move this into a class of its own?
+    public static class OutOfRangeSpec {
```
Feedback appreciated on this one

shans96 commented on x-pack/plugin/mapper-unsigned-long/src/test/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongFieldMapperTests.java:
```diff
@@ -365,6 +370,38 @@ protected IngestScriptSupport ingestScriptSupport() {
+        return 123L;
+    }
+
+    @Override
+    protected Number randomNumber() {
+        if (randomBoolean()) {
+            return randomLong();
+        }
+        if (randomBoolean()) {
+            return randomDouble();
+        }
+        assumeFalse("https://github.com/elastic/elasticsearch/issues/70585", true);
+        // TODO: unsigned long has a max value of (2^64)-1, but it can't be used in the randomDoubleBetween method because it's too large
+        return randomDoubleBetween(0L, Long.MAX_VALUE, true);
+    }
```
Still need to resolve this one as well, was thinking of making a new method that supports larger numbers but maybe there's an easier way to support this for unsigned longs?

shans96 commented:
Hi @cbuescher, thanks for the feedback. I've made changes based on your comments and think the PR is ready for a proper review; `UnsignedLongFieldMapperTests` now extends `WholeNumberFieldMapperTests` and `ScaledFloatFieldMapperTests` now extends `NumberFieldMapperTests`. I had some small questions on making the code cleaner where I've overridden the methods, and some other small questions, but hopefully they are quick to resolve :+1: 

Look forward to hearing from you!

cbuescher reviewed (changes requested):
Sorry for the late reply. I left a few comments but nothing big, I think we should merge this and work on extending the tests to fill in the missing methods later. As it is, this already improves things by using the same base test class and doesn't remove any coverage. Can you update the branch and add the few things I suggested before I run our CI test suite please?

shans96 committed:
Delete TODO

cbuescher commented:
@shans96 just checking in to see if you want to continue this PR, since the branch has gone stale a bit and would need updating. As I mentioned earlier in the comments I think there are only a few smaller changes left to address before we can merge this.

shans96 committed:
Add comments based on feedback

shans96 commented:
Hi @cbuescher, apologies for the delay in reply on this one. Keen to finish this PR and will make it a focus- I've updated the comments based on your feedback. I'm currently looking at making `OutOfRangeSpec` publicly accessible: adding a dependency on `elasticsearch.server.test` wouldn't make sense (and would cause a circular dependency I think). Thinking of moving the class to the test framework instead- would this make sense?

cbuescher commented:
> Thinking of moving the class to the test framework instead- would this make sense

It think so, yes. Maybe also make the name a bit more specific then, since that class seems tied to the number type tests, so maybe "NumberTypeOutOfRangeSpec" or something along those lines would be good. Please also try to update the branch, I can help with that if needed.

cbuescher commented:
Hi @shans96, I'll hand this over to @pmpailis while I'm out for a few weeks. I we only need updating the branch and move the shared "OutOfRangeSpec" code somewhere where it can be accessed by modules. That is, if all tests pass in CI.

shans96 committed:
Move OutOfRangeSpec to the test framework under the name 'NumberTypeOutOfRangeSpec'

shans96 committed:
Merge branch 'master' into share_field_mapper_tests

shans96 commented:
Hi, apologies again for the delay in updates on this one- it's been a hectic holiday season. I've moved `OutOfRangeSpec` into the test framework under the name `NumberTypeOutOfRangeSpec`- not sure if the class location/name are ideal, so any feedback on this one would be appreciated. I've also updated this branch with the latest changes from `master` and everything builds successfully on my local machine, so I think it's ready for CI tests.

pmpailis commented:
Thanks for continuing this work @shans96 - changes LGTM! 
The only thing pending prior to running CI is to address the conflicts with `main` as some of the files above were changed recently (e.g. [here](https://github.com/elastic/elasticsearch/pull/103510/files#diff-1a110e1edc49686c15688a028b756a0a4d3cebffae59e5e3d85ae3e0ca76cf37R370)). Happy to help with addressing this if needed :)

pmpailis commented:
@shans96 happy to assist with merging/resolving conflicts to move this forward :)

shans96 committed:
Merge branch 'master' into share_field_mapper_tests

shans96 commented:
Hi, thanks for the ping :+1: I've just merged the latest version of `main` into my branch, but haven't run into any merge conflicts. I was anticipating some, so I looked at the PR you linked and tried to search for the new method declarations in the latest version of `main`, but they don't seem to exist there either. Wondering if I missed something, or were there additional changes afterwards that kept the field mapper tests as they were originally?

pmpailis commented:
Thanks for taking a look @shans96 ! Yeah, actually it seems that some changes were reverted in [a later PR](https://github.com/elastic/elasticsearch/pull/104218/files#diff-031008a5bee59ebc302cdd947cd0259f5aa716cb49e4a69efd73cd968956b85dR385) a few days ago, so all seems fine now :)  Will proceed with running CI tests :)

pmpailis commented:
@elasticmachine test this please

pmpailis commented:
buildkite test this please

pmpailis commented:
Thank you for your effort & contribution @shans96 ! CI tests have all passed 🚀  so I'll proceed with merging :)
